File: payment-service/src/main/java/pl/altkom/asc/lab/micronaut/poc/payment/domain/PolicyUnregisteredListener.java
package pl.altkom.asc.lab.micronaut.poc.payment.domain;

import pl.altkom.asc.lab.micronaut.poc.policy.service.api.v1.events.PolicyUnregisteredEvent;

import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.OffsetReset;
import io.micronaut.configuration.kafka.annotation.Topic;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@KafkaListener(offsetReset = OffsetReset.EARLIEST)
public class PolicyUnregisteredListener {
    private final PolicyAccountRepository policyAccountRepository;

    @Topic("policy-unregistered")
    void onPolicyUnregistered(PolicyUnregisteredEvent event) {
        policyAccountRepository.delete(event.getPolicyAccountId());
    }
}

File: documents-service/src/main/kotlin/pl/altkom/asc/lab/micronaut/poc/documents/domain/PolicyUnregisteredListener.kt
package pl.altkom.asc.lab.micronaut.poc.documents.domain

import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.OffsetReset
import io.micronaut.configuration.kafka.annotation.Topic
import io.micronaut.http.annotation.Header
import pl.altkom.asc.lab.micronaut.poc.policy.service.api.v1.events.PolicyUnregisteredEvent

@KafkaListener(clientId = "policy-unregistered-listener", offsetReset = OffsetReset.EARLIEST)
@Header(name = "Content-Type", value = "application/json")
class PolicyUnregisteredListener(private val policyDocumentRepository: PolicyDocumentRepository) {

    @Topic("policy-unregistered")
    fun onPolicyUnregistered(event: PolicyUnregisteredEvent) {
        policyDocumentRepository.delete(event.policyDocumentId)
    }
}
